dubbo-rpc/dubbo-rpc-dubbo/src/main/java/org/apache/dubbo/rpc/protocol/dubbo/telnet/ListTelnetHandler.java

```diff
@@ -46,55 +51,107 @@ public String telnet(Channel channel, String message) {
-                    buf.append("\r\n");
-                }
-                buf.append(exporter.getInvoker().getInterface().getName());
-                if (detail) {
-                    buf.append(" -> ");
-                    buf.append(exporter.getInvoker().getUrl());
-                }
-            }
+            printAllServices(buf, detail);
         } else {
-            Invoker<?> invoker = null;
-            for (Exporter<?> exporter : DubboProtocol.getDubboProtocol().getExporters()) {
-                if (service.equals(exporter.getInvoker().getInterface().getSimpleName())
-                        || service.equals(exporter.getInvoker().getInterface().getName())
-                        || service.equals(exporter.getInvoker().getUrl().getPath())) {
-                    invoker = exporter.getInvoker();
-                    break;
+            printSpecifiedService(service, buf, detail);
+
+            if (buf.length() == 0) {
+                buf.append("No such service: ").append(service);
+            }
+        }
+        return buf.toString();
+    }
+
+    private void printAllServices(StringBuilder buf, boolean detail) {
+        printAllProvidedServices(buf, detail);
+        printAllReferredServices(buf, detail);
+    }
+
+    private void printAllProvidedServices(StringBuilder buf, boolean detail) {
+        boolean printHeader = true;
+        for (ProviderModel provider : ApplicationModel.allProviderModels()) {
+            if (printHeader) {
+                buf.append("PROVIDER:\r\n");
+                printHeader = false;
+            }
+
+            buf.append(provider.getServiceName());
+            if (detail) {
+                buf.append(" -> ");
+                buf.append(" published: ");
+                buf.append(isRegistered(provider.getServiceName()) ? "Y" : "N");
+            }
+            buf.append("\r\n");
+        }
+    }
+
+    private void printAllReferredServices(StringBuilder buf, boolean detail) {
+        boolean printHeader = true;
+        for (ConsumerModel consumer : ApplicationModel.allConsumerModels()) {
+            if (printHeader) {
+                buf.append("CONSUMER:\r\n");
+                printHeader = false;
+            }
+
+            buf.append(consumer.getServiceName());
+            if (detail) {
+                buf.append(" -> ");
+                buf.append(" addresses: ");
+                buf.append(getConsumerAddressNum(consumer.getServiceName()));
+            }
+        }
+    }
+
+    private void printSpecifiedService(String service, StringBuilder buf, boolean detail) {
+        printSpecifiedProvidedService(service, buf, detail);
+        printSpecifiedReferredService(service, buf, detail);
+    }
+
+    private void printSpecifiedProvidedService(String service, StringBuilder buf, boolean detail) {
+        for (ProviderModel provider : ApplicationModel.allProviderModels()) {
+            if (service.equalsIgnoreCase(provider.getServiceName())
```

Comment:
can be change to same above one of the example
`if(isServiceMatch(service,provider)) {
buf.append(provider.getServiceName()).append(" (as provider):\r\n");
for (ProviderMethodModel method : provider.getAllMethods()) {
printMethod(method.getMethod(), buf, detail);
}
}`

Reply:
done.